### PR TITLE
remove stock icons

### LINKED ISSUE #84: Move away from using stock icons
Stock icons are deprecated in GTK and many custom icon themes do not provide them anymore so TRG ends up using ugly icons from the default system theme.

### LINKED ISSUE #123: better distinguish buttons "verify" and "re-announce"
I have repeatedly mistakenly clicked "verify" when I instead wanted to click "re-announce".

Two reasons make these very easy to mix up after right-clicking on a torrent:

- Both have the same icon
image
- Both are displayed together, "re-announce" immediately below "verify":
https://github.com/transmission-remote-gtk/transmission-remote-gtk/blob/dff54e8df6d2961495f94f9291c33455c6068bbb/src/trg-menu-bar.c#L48-L49

I propose to:

1. Create a new icon for one of them.
2. Re-order the list to not have them right after another.

(Alternatively, I would also like an option to disable displaying "re-announce" on right-click, leaving it only in the "Torrent" main toolbar menu.)

Thanks!

### PR DESCRIPTION
**Pull request by LaserEyess**
I'm just opening this for my own book keeping of things that are still broken. This is in no way complete and should not be merged. I may need help fixing the rest of it

 - [x] Remove all instances of `GTK_STOCK_*` and deprecated functions containing `_stock`
 - [x] Figure out what to do for all `GTK_STOCK_*` that do not have named icons or labels replacements, e.g. `gtk-connect`
 - [x] ~~Should `trg_imagemenuitem_box()` be split into it's own commit?~~ no, will need further refactoring anyway
 - [x] In `trg_state_selector_consturctor()` in `trg-state-selector.c`, `gtk_list_store_new()` needs to be updated to use the correct icon type. No idea how to do this, the docs are miserable
 - [x] Fix #123 because why not
 - [x] Split out `gtk-connect`, `gtk-disconnect`, `gtk-apply` into a gresource.xml and add it to the source

Closes #84
Closes #123

### CODE REVIEW

**Comment by TingPing** [comment]
> In trg_state_selector_consturctor() in trg-state-selector.c, gtk_list_store_new() needs to be updated to use the correct icon type. No idea how to do this, the docs are miserable

Its `icon-name` instead of `stock-id`: 
https://developer-old.gnome.org/gtk3/stable/GtkCellRendererPixbuf.html#GtkCellRendererPixbuf--icon-name

**Comment by LaserEyess** [comment]
Wow that worked instantly, thank you.

**Comment by LaserEyess** [comment]
So it seems like menu icons are likely impossible. Previously, there was `GtkImageMenu` which allowed you to just pack an image. That's now deprecated. I found a stack overflow answer about re-implementing it https://stackoverflow.com/a/45211700, key text:

> However that only works if your menu does not have a GtkCheckMenuItem in it, since the deprecated GtkImageMenuItem had a hack in it to use the pre-reserved toggle size. I could not find a way to get a similar result to the original hack, which btw. was annotadet by the gtk team with:
> 
>  ```
>  /* Man this is lame hardcoding action, but I can't
>  * come up with a solution that's really better.
>  */
> ```
> 

We have multiple menus with `GtkCheckMenuItem` and the like. This is in line with what was said in #117 about the issue. I will not spend the time doing this, at least. If someone in the future is looking at this and seething about the menu icons missing, I'm sorry, but you'll need to take it up with GTK devs. They're the ones who removed this without a replacement.

**Comment by TingPing** [comment]
I wouldn't be surprised if the GTK3 bug they mentioned no longer applies to GTK4.

**Comment by LaserEyess** [comment]
@TingPing would appreciate a review if you have time. I know I still have some TODO's littered around the code, but I don't actually know if I can replace things like `gtk-connect` with more appropriate icon names.

The only major change between this commit and git master would be that menus in the menu bar no longer have icons. Other menus do, however.

**Comment by TingPing** [comment]
> but I don't actually know if I can replace things like `gtk-connect` with more appropriate icon names.

The immediate solution is copy gtk-connect icon into this project.

**Comment by TingPing** [comment]
This change looks fine to me. Again I don't think the menu icons are really valuable so that's OK.

I would probably even use the symbolic versions in most places but that can be a future discussion.

**Comment by LaserEyess** [comment]
> The immediate solution is copy the old icon into this project.


As long as this project stays on GTK3 I think that won't be necessary. I think someone just needs to pick what the new ones should be. There are 3:
- `gtk-connect`
- `gtk-apply`
- `gtk-disconnect`

Hypothetically anything in [this list](https://specifications.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.html) should work? I don't see anywhere in the GTK4 docs where they actually define what valid icon names are, though that could be me not understanding where to look.

**Comment by LaserEyess** [comment]
Missed a couple `*_from_stock` functions as well as some references to "stock" that were not correct, changed to "label" instead.

**Comment by TingPing** [comment]
> Hypothetically anything in this list should work? I don't see anywhere in the GTK4 docs where they actually define what valid icon names are, though that could be me not understanding where to look.

Icons are not a part of GTK. They are simply system data following the icon-theme spec.

So how it works is you can safely assume everybody has `hicolor-icon-theme`, and well, thats kinda all. Many GNOME users will have `adwaita-icon-theme` but you can't actually rely on it. The modern idea is you bundle icons you want to use because that will simply always work.

**Comment by LaserEyess** [comment]
Thanks again for the tips, I'll look into this. I really appreciate all the help (handholding) you've been doing. The build system changes are going to be, uh, tricky... but I have some autotools experience so I should be able to hack it. Honestly changing the build system would be a huge benefit to this project, too. 

I don't think there's a huge risk of merge conflicts arising so I think I can just tack those changes on as another commit.

**Comment by TingPing** [comment]
> The build system changes are going to be, uh, tricky... but I have some autotools experience so I should be able to hack it

Here is an example of using gresources in autotools: https://github.com/TingPing/transmission-remote-gnome/blob/9b947e0a6b8a86fb3353536226df13e02aa96f60/Makefile.am#L30-L36

> Honestly changing the build system would be a huge benefit to this project, too.

I'm all for `meson` if you want to do that. I'll gladly review it.

**Comment by LaserEyess** [comment]
FYI new force pushes are some warnings I fixed from code I wrote. Basically I forgot to turn on regular warnings after checking for deprecations. Things like: 
```
trg-status-bar.c: In function ‘turtle_toggle’:
trg-status-bar.c:91:59: warning: passing argument 2 of ‘gtk_image_get_icon_name’ from incompatible pointer type [-Wincompatible-pointer-types]
   91 |     gtk_image_get_icon_name(GTK_IMAGE(priv->turtleImage), &iconName, NULL);
      |                                                           ^~~~~~~~~
      |                                                           |
      |                                                           gchar ** {aka char **}
In file included from /usr/include/gtk-3.0/gtk/gtkentry.h:42,
                 from /usr/include/gtk-3.0/gtk/gtktreeview.h:29,
                 from /usr/include/gtk-3.0/gtk/gtkcombobox.h:27,
                 from /usr/include/gtk-3.0/gtk/gtkappchooserbutton.h:29,
                 from /usr/include/gtk-3.0/gtk/gtk.h:42,
                 from trg-status-bar.c:26:
/usr/include/gtk-3.0/gtk/gtkimage.h:192:51: note: expected ‘const gchar **’ {aka ‘const char **’} but argument is of type ‘gchar **’ {aka ‘char **’}
  192 |                                     const gchar **icon_name,
      |                                     ~~~~~~~~~~~~~~^~~~~~~~~
trg-menu-bar.c: In function ‘trg_menu_bar_item_new’:
trg-menu-bar.c:343:37: warning: passing argument 1 of ‘gtk_menu_item_set_use_underline’ from incompatible pointer type [-Wincompatible-pointer-types]
  343 |     gtk_menu_item_set_use_underline(item, TRUE);
      |                                     ^~~~
      |                                     |
      |                                     GtkWidget * {aka struct _GtkWidget *}
In file included from /usr/include/gtk-3.0/gtk/gtkcheckmenuitem.h:33,
                 from /usr/include/gtk-3.0/gtk/gtk.h:72,
                 from trg-menu-bar.c:25:
/usr/include/gtk-3.0/gtk/gtkmenuitem.h:154:69: note: expected ‘GtkMenuItem *’ {aka ‘struct _GtkMenuItem *’} but argument is of type ‘GtkWidget *’ {aka ‘struct _GtkWidget *’}
  154 | void       gtk_menu_item_set_use_underline    (GtkMenuItem         *menu_item,
      |

**Comment by LaserEyess** [comment]
@TingPing so, I think I have the makefile changes made, that was not the hard part I guess. Now I have three major questions

First question: looking at the docs for [Gio.Resource](https://docs.gtk.org/gio/struct.Resource.html) and [Gtk.Application](https://docs.gtk.org/gtk3/class.Application.html) it seems like it will somehow automagically find these icons and I can just sue `"gtk-connect"` (or whatever) as a valid icon name in the code. However:

> GtkApplication will also automatically setup an icon search path for the default icon theme by appending “icons” to the resource base path. This allows your application to easily store its icons as resources. See gtk_icon_theme_add_resource_path() for more information.

Since `gtk_icon_theme_add_resource_path()` has absolutely nothing about this, what is the resource base path? Where is it documented? I'm assuming something like `/io/github/TransmissionRemoteGtk/` or something but I can't seem to find it. Any tips would be appreciated.

Second question: Do I have to nest every single icon at every single size inside of the XML? If so, is there a way to automate that?

Third question: where do I get the icon images? I have no idea what license they are, either.

File: src/Makefile.am [hunk 1]
```diff
@@ -17,6 +17,21 @@
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
 
+
+gresource_xml  := $(top_srcdir)/data/io.github.TransmissionRemoteGtk.gresource.xml
```
Comment (TingPing) [inline src/Makefile.am:21]: I believe this should be added to `EXTRA_DIST` (as in `make distcheck` probably fails atm).

**Comment by TingPing** [comment]
> what is the resource base path? Where is it documented? I'm assuming something like /io/github/TransmissionRemoteGtk/

You are correct. It is documented [here](https://docs.gtk.org/gio/method.Application.set_resource_base_path.html).

> By default, the resource base path is determined from the application ID by prefixing ‘/’ and replacing each ‘.’ with ‘/’. This is done at the time that the GApplication object is constructed. Changes to the application ID after that point will not have an impact on the resource base path.

**Comment by TingPing** [comment]
> Second question: Do I have to nest every single icon at every single size inside of the XML? If so, is there a way to automate that?

We probably only use them at a single size. You can use the `alias` option to rename where they are located in the resource vs on disk. Like [this](https://github.com/TingPing/transmission-remote-gnome/blob/master/data/se.tingping.Trg.gresource.xml#L17).

> Third question: where do I get the icon images? I have no idea what license they are, either.

- https://github.com/GNOME/gtk/tree/gtk-3-24/gtk/icons
- https://github.com/GNOME/gtk/blob/gtk-3-24/COPYING

LGPL-2.0

**Comment by LaserEyess** [comment]
Hm, didn't even think to check gtk's tree for them, that's way smarter than `find` on `/usr/share/icons`. Thanks again for the hints, I think this all makes sense now.

File: src/Makefile.am (hunk 1 above)
Comment (LaserEyess) [reply src/Makefile.am:21]: Good catch, RESOURCE_FILES as well

**Comment by TingPing** [comment]
Ready for final review/merge?

**Comment by LaserEyess** [comment]
Almost, it looks like most projects keep their `*gresource.xml` in their `src/` or equivalent tree, I guess that makes sense because of generating their resource files there as well. I was going to make that change as well. Furthermore, I realized trg [bundles the turtle icons](https://github.com/transmission-remote-gtk/transmission-remote-gtk/blob/master/src/icon-turtle.h) separately. I think putting them in the same place as the other icons makes more sense, and it allows this project to drop a couple of files out of the source tree. 

I was going to finish that tonight and then ping you, so if you have time tomorrow or sometime this weekend to review then I'd appreciate it.

**Comment by LaserEyess** [comment]
Decided against moving gresouces.xml, I feel like it makes way more sense in `data/` even if that may not be the convention others use. I did, however, rename the icon paths and use `alias=` as suggested earlier. That way makes it much more obvious what icons are being used for what I think. Plus it removes some empty dirs in the repo. Feel free to bikeshed, I have already `git rebase -i`'d multiple times anyway. 

@TingPing I think I'm ready now. I appreciate your patience these last few days, I've learned a lot about gtk so hopefully more changes will come with less questions in the future.

**Comment by LaserEyess** [comment]
Oh, one final thing I think I should mention. For #123, I changed the icon for "reannounce" in the context menu to be `network-server` instead of `view-refresh`, to distinguish it from "verify". That is pretty much the only other cosmetic difference between current master and this PR, besides the menu icons being gone.

**Comment by TingPing** [comment]
@LaserEyess You did great, thanks for the contribution :)

**Pull request merged**